food-truck-api/src/main/java/food/truck/api/endpoint/ReviewEndpoint.java

```diff
@@ -0,0 +1,94 @@
+import org.springframework.lang.Nullable;
+import org.springframework.security.access.annotation.Secured;
+import org.springframework.security.core.annotation.AuthenticationPrincipal;
+import org.springframework.web.bind.annotation.*;
+import org.springframework.web.server.ResponseStatusException;
+
+import java.util.LinkedList;
+import java.util.List;
+import java.util.Optional;
+
+@Log4j2
+@RestController
+public class ReviewEndpoint {
+    @Autowired
+    private ReviewService reviewService;
+
+    @Autowired
+    private UserService userService;
+
+    @Autowired
+    private TruckService truckService;
+
+    @GetMapping("/reviews/{reviewId}")
+    public Review getReviewById(@PathVariable long reviewId) {
+        return reviewService.findReviewById(reviewId);
+    }
+
+    @GetMapping("/reviews/user")
+    public List<Review> getUserReviews(@RequestParam String username) {
+        User user = userService.loadUserByUsername(username);
+        if(user == null){
+            return new LinkedList<Review>();
+        }
+        return reviewService.findReviewsByUserId(user.getId());
+    }
+
+    @GetMapping("/reviews/user/{userId}")
+    public List<Review> getUserReviews(@PathVariable long userId) {
+        return reviewService.findReviewsByUserId(userId);
+    }
+
+    @GetMapping("/reviews/truck/{truckId}")
+    public List<Review> getTruckReviews(@PathVariable long truckId) {
+        return reviewService.findReviewByTruckId(truckId);
+    }
+
+    @GetMapping("/reviews/truck/{truckId}/user")
+    public Review getTruckReviews(@PathVariable long truckId, @RequestParam Long userId) {
+        return reviewService.findReviewForUser(userId, truckId);
+    }
+
+    @Value
+    private static class PostReviewParams {
+        Long userId;
+        int score;
+        int costRating;
+        @Nullable
+        String reviewText;
+    }
+
+    @Secured({"ROLE_USER"})
+    @PostMapping("/reviews/truck/{truckId}")
+    public Review postTruckReview(@AuthenticationPrincipal User user, @PathVariable long truckId, @RequestBody PostReviewParams data) {
+        // TODO decide if we want to just silently change the review to match
+        if (!user.getId().equals(data.userId))
+            throw new ResponseStatusException(HttpStatus.FORBIDDEN, "UserID does not match userID of the provided review");
+
+        // Ensure truck exists
+        Optional<Truck> truck = truckService.findTruckById(truckId);
+        if (truck.isEmpty())
+            throw new ResponseStatusException(HttpStatus.NOT_FOUND, "Provided truck does not exist");
+
+        // Save review
+        return reviewService.saveReview(data.userId, truck.get(), data.score, data.costRating, data.reviewText);
+    }
+
+    @Secured({"ROLE_USER"})
+    @DeleteMapping("/reviews/truck/{truckId}")
+    public boolean deleteReview(@AuthenticationPrincipal User u, @PathVariable long truckId) {
+        return reviewService.deleteReviewByUser(u, truckId);
```

Comment:
These should be checked that the review belongs to the user

Reply:
Authentication; what a concept! Will fix.

Reply:
Forgot to follow up on this: so the way it fetches the review in the service, it actually doesn't need/can't be authenticated. It fetches the review matching the userId from AuthenticationPrinciple and the provided truckId, so by definition the user can't delete someone else's review unless they're signed in as someone else or defeat our authentication system, neither of which we can't check for.